src/las/point.cpp

```diff
@@ -0,0 +1,228 @@
+        internal::pack(user_data_, out_stream);
+        internal::pack(extended_scan_angle_, out_stream);
+    }
+    else
+    {
+        internal::pack(returns_flags_eof_, out_stream);
+        internal::pack(classification_, out_stream);
+        internal::pack(scan_angle_rank_, out_stream);
+        internal::pack(user_data_, out_stream);
+    }
+    internal::pack(point_source_id_, out_stream);
+
+    if (has_gps_time_)
+        internal::pack(gps_time_, out_stream);
+    if (has_rgb_)
+    {
+        internal::pack(rgb_[0], out_stream);
+        internal::pack(rgb_[1], out_stream);
+        internal::pack(rgb_[2], out_stream);
+    }
+    if (has_nir_)
+        internal::pack(nir_, out_stream);
+
+    for (auto eb : extra_bytes_)
+        internal::pack(eb, out_stream);
+}
+
+uint8_t Point::BaseByteSize(const uint8_t &point_format_id)
+{
+    switch (point_format_id)
+    {
+    case 0:
+        return 20;
+    case 1:
+        return 28;
+    case 2:
+        return 26;
+    case 3:
+        return 34;
+    case 4:
+        return 28;
+    case 5:
+        return 34;
+    case 6:
+        return 30;
+    case 7:
+        return 36;
+    case 8:
+        return 38;
+    case 9:
+        return 30;
+    case 10:
+        return 38;
+    default:
+        throw std::runtime_error("Point format must be 0-10");
+        ;
+    }
+}
+
+void Point::ToPointFormat(const uint8_t &point_format_id)
+{
+
+    if (extended_point_type_ && point_format_id < 6)
+    {
+        // Convert points values from 1.4 to 1.0
+        // Returns and flags
+        returns_flags_eof_ = (EdgeOfFlightLineFlag() << 7) | (ScanDirectionFlag() << 6) |
+                             (std::min(int(NumberOfReturns()), 7) << 3) | std::min(int(ReturnNumber()), 7);
+        // Classification
+        classification_ =
+            (Withheld() << 7) | (KeyPoint() << 6) | (Synthetic() << 5) | std::min(int(Classification()), 31);
+
+        // Scan angle
+        scan_angle_rank_ = static_cast<int8_t>(std::max(-90, std::min(90, int(float(ExtendedScanAngle()) * 0.006f))));
+    }
+    else if (!extended_point_type_ && point_format_id > 5)
+    {
+        // Convert point values from 1.0 to 1.4
+        // Returns
+        extended_returns_ = (NumberOfReturns() << 4) | ReturnNumber();
+        // Flags
+        extended_flags_ = (EdgeOfFlightLineFlag() << 7) | (ScanDirectionFlag() << 6) | (Withheld() << 2) |
+                          (KeyPoint() << 1) | Synthetic();
+        // Classification
+        extended_classification_ = Classification();
+        // Scan angle
+        extended_scan_angle_ = static_cast<int16_t>(float(ScanAngleRank()) / 0.006f);
+    }
+
+    // Update flags
+    extended_point_type_ = point_format_id > 5;
+    has_gps_time_ = FormatHasGpsTime(point_format_id);
+    has_rgb_ = FormatHasRgb(point_format_id);
+    has_nir_ = FormatHasNir(point_format_id);
+    point_byte_size_ = BaseByteSize(point_format_id);
```

Comment:
`+ len(extended bytes)`

Reply:
Should the point_record_length_ include the size of extra bytes?

Reply:
I currently just have it as the minimum point size, and the number of extra bytes can be accessed by calling ExtraBytes().size()

Reply:
I'm also planning to change from vector to char* for extra bytes, I think the vectors may take more memory when we don't really need their features.

Reply:
In which case I can have point_record_length_ as the length of the point data, and either a variable storing the number of extra bytes, or a variable storing the total byte size of point+EBs. What do you think?

Reply:
I think point_record_length_  should include the extra byte size, to make it consistent with the LAS spec. What lazperf does internally is stores the point_record_length inluding the extra bytes, then on functions where it needs to know the original point size, you pass in the point_record_id and the number of extra bytes (computed by point_record_length - BaseSize(point_record_id). 

I think the vectors have pretty low overhead and are "safer" than char* objects. I think it's good to err on the side of c++ oriented code, trying to avoid C features. But let me know what you think.

Reply:
Cool, I don't really like char* either haha. Wasn't sure how much overhead vectors added. Happy to keep 'em!